Template sensor attributes

gadgetchnnel opened the pull request:
## Description:
Added option to add custom attributes to the template sensor and binary sensor via a new **attribute_templates** config option.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io/pull/8278

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: template
    sensors:
      test_sensor:
        value_template: >-
          {{ states.device_tracker.my_phone.state == "home" }}
        attribute_templates:
          latitude: >-
            {{ states.device_tracker.my_phone.attributes['latitude'] }}
          longitude: >-
            {{ states.device_tracker.my_phone.attributes['longitude'] }}

sensor:
  - platform: template
    sensors:
      test_sensor:
        value_template: >-
          {{ states.device_tracker.my_phone.state }}
        attribute_templates:
          latitude: >-
            {{ states.device_tracker.my_phone.attributes['latitude'] }}
          longitude: >-
            {{ states.device_tracker.my_phone.attributes['longitude'] }}
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

gadgetchnnel committed:
template sensor and binary_sensor attributes

Added support for custom attributes to the template sensor and binary_sensor

gadgetchnnel committed:
Fix duplicated line

Fix duplicated line

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-sensor-attributes

gadgetchnnel committed:
Fixed error message logic and added tests

Fixed the logic which removes_template from name to work with new attribute_templates and added tests.

gadgetchnnel committed:
Fix to new template binary_sensor test

Fix to new template binary_sensor tests

gadgetchnnel committed:
Fixed tests

Fixed tests

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-sensor-attributes

homeassistant commented:
Hi @gadgetchnnel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#20410).

Once you do that we will be able to review and accept this pull request.

Thanks!

gadgetchnnel committed:
Fixed whitespace and line length issues

Fixed whitespace and line length issues

homeassistant commented:
Hi @gadgetchnnel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#20410).

Once you do that we will be able to review and accept this pull request.

Thanks!

gadgetchnnel committed:
More whitespace and line length issues fixed

More whitespace and line length issues fixed

gadgetchnnel committed:
Indentation fixes

Indentation fixes

gadgetchnnel committed:
Line length fixes

Line length fixes

gadgetchnnel committed:
Updated indentation

Updated indentation to fix pylint errors

gadgetchnnel committed:
moved to new folder structure

moved to new folder structure

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-sensor-attributes

gadgetchnnel committed:
Revert "Merge remote-tracking branch 'upstream/dev' into template-sensor-attributes"

This reverts commit 13e6f60838352d3a1d3c493a88c79f92525c5244, reversing
changes made to b8efad61551d9b20d14e2fc1164d1835e4f6678c.

gadgetchnnel committed:
Revert merge

awarecan committed:
Fix tts Great Migration issue (#22539)

gadgetchnnel committed:
Fix binary_sensor test

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-sensor-attributes

Yarikx committed:
Ring camera improvements (#22526)

* Ring camera improvements

Expose last_video_id attribute.
Fix missing last_video_url
Only update last_video_id when video is ready

* Fix formatting

ktnrg45 committed:
PS4 bump to 0.5.2 (#22523)

* Bump pyps4 to 0.5.2

* Bump pyps4 to 0.5.2

* Bump pyps4 to 0.5.2

gadgetchnnel committed:
Fixed mock.patch

dmulcahey committed:
clean up channel configuration (#22534)

syssi committed:
Fix regression of the xiaomi_aqara config validation (#22435)

* Fix regression of the xiaomi_aqara config validation

* Make the key optional again

* Add base schema

* Remove the GW_MAC default

StevenLooman committed:
Upgrade to async_upnp_client==0.14.7 (#22543)

damarco committed:
Bump zigpy (#22545)

awarecan committed:
Set up Circleci workflow (#22519)

* Set up Circleci workflow

* Update python tag

* Add pre-test job to cache the requirements

* Upgrade pip itself

* Use 3.7 for lint

* Parallelize pylint

* Tweak run gen_requirements_all

* tweak cache key

balloob committed:
Updated frontend to 20190329.0

balloob committed:
Add support for streaming to ffmpeg (#22549)

balloob committed:
Fix platform warnings (#22551)

royto committed:
upgrade pylinky to 0.3.3 (#22544)

balloob committed:
Move core services.yaml file to Home Assistant integration (#22489)

* Move services.yaml to correct dir

* Remove special case for HA servicesgs

giefca committed:
Google assistant: add blinds trait for covers (#22336)

* Update const.py

* Update smart_home.py

* Update trait.py

* Update test_trait.py

* Update smart_home.py

* Update test_trait.py

* Update trait.py

* Update trait.py

* Update test_trait.py

* Update test_trait.py

* Update __init__.py

* Update test_trait.py

* Change email

* Trying to correct CLA

* Update __init__.py

* Update trait.py

* Update trait.py

* Update trait.py

* Update trait.py

* Update __init__.py

* Update test_trait.py

* Update test_google_assistant.py

* Update trait.py

* Update trait.py

* Update test_trait.py

* Update test_trait.py

drjared88 committed:
Update Amcrest component to SUPPORT_STREAM (#22553)

* Update camera.py

Update Amcrest component to SUPPORT_STREAM to allow streaming in the UI and Google Assistant.

* Update camera.py

andrewsayre committed:
Change HEOS component library and add basic config flow (#22517)

* Update heos lib

* Update requirements files

* Removed unecessary mock_coro usage

* Remove assert_called_once usage

* Updates from review feedback

* Remove extra param to error format

Commit:
Add command_template and value_template for MQTT alarm (#21438)

* Option to send pin code with the MQTT payload for MQTT alarm

* publish code via json

Add publish code via json
add code_disarm_required

* publish code via json

Add publish code via json
add code_disarm_required

* implemented command_template

* Fix issue with night arm and add template test

* implemented value_template for mqtt alarm

* Fixed merge errors

* Requested changes

* Resolve lint errors

* Resolve hound issues

* Fix test formatting

amelchio committed:
Add table with netgear_lte sensor units (#22508)

Jc2k committed:
Add more HomeKit device enumeration tests (#22194)

* Test that Aqara Gateway, Ecobee 3 and Lennox E30 is correctly enumerated

* Move json to fixtures directory

* Move IO to executor

Oro committed:
Added support for transitions for nanoleaf light (#22192)

* Added transition support for nanoleaf

* Formatting for comments

* Inline comment instead of additional line

* Set color_temp before starting transition

andrewsayre committed:
Add Heos config flow (#22554)

* Add UI initiated config flow

* Fix alpha order

awarecan committed:
Fix name conflict in tests (#22556)

* Fix name conflict in tests

* Lint

* Lint

emontnemery committed:
Improve handling of audio groups (#22396)

* Improve handling of audio groups

* Review comments

* Fix tests

* Add tests

* Review comment

* Bump pychromecast

carstenschroeder committed:
Prevent toogle to false at restart of ADS platforms (#22522)

* Prevent toogle to false at restart

* change to asyncio.run_coroutine_threadsafe

kstaniek committed:
Ampio Smog Air Quality Sensor (#21152)

* Initial commit for Ampio Smog Air Quality Sensor

* coveragerc and requirements_all update

* Lint fixed

* Moved to vendor folder

* Updated according to review

* Docs string fix

* Docstring fix

* Docstring fix

* Requirements fixed

* Lint fix

* .coveragerc updated

pavoni committed:
Bump pyloopenergy library to 0.1.2 (#22561)

balloob committed:
Delete azure-pipelines.yml

pvizeli committed:
Delete main.workflow

emontnemery committed:
Forward media control to playing group (#22566)

* Forward media control to playing group

* Fix forwarding control to dynamic group

* Fix, add tests

CWhits committed:
Update Foscam component to support stream source (#22568)

* Update Foscam to support stream source

* Removing spaces and tabs

* Changing to Python3-style string formatting

* Adding '_media_port' to hopefully cover other models

emontnemery committed:
Turn light off if brightness is 0 (#22400)

* Turn light off if brightness is 0

* Lint

* Review comments

* Lint

* Fixup, add tests

* Fix trådfri light + test

awarecan committed:
Ignore flaky test (#22563)

OleksandrBerchenko committed:
Fix typo in light/__init__.py (#22581)

amelchio committed:
Fix lightwave config validation (#22576)

OleksandrBerchenko committed:
Rewrite Osram Lightify component (#22184)

* Rewrite Osram Lightify component

* Update python-lightify version to 1.0.7.2

* Remove unneeded code

* 1. Remove changes in light/__init__.py, 2. Set properties to None by default

* Fix typo

Commit:
Clean up homematicip cloud (#22589)

* Code Cleanup

- removed unused constants
- more icons on binary_sensor groups
- alligned  code for device_state_attributes
- fixed temperature unit origin for weather

* removed icons

Commit:
Retrying connecting Influxdb at setup (#22567)

* Also retry Influxdb at setup()

* Use event.call_later() for retry setup Influxdb

* Fix max line length in setup() in Influxdb

* Add extra space before comment

* Fix sec -> seconds and add return True

renemarc committed:
Fix gtfs typing and logger issues (#22572)

## Description:
Some code cleanup requests where raised in the [latest merged GTFS commit](https://github.com/home-assistant/home-assistant/pull/20966/commits/9153e3b671990d3c33f59b8cde5506b30fcaaa65). This new PR aims to address them, including:
- Clear all typing issues.
- Respect logger levels and format.
- Simplify some non-pythonic lines.

This sensor now passes `mypy` testing, but does so by ignoring two lines with `# type: ignore`.

**Related issue (if applicable):** fixes issues raised by @MartinHjelmare in #20966

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

drjared88 committed:
Update ONVIF component to SUPPORT_STREAM (#22569)

* Update Onvif component to SUPPORT_STREAM

* Update camera.py

* Update camera.py

* Update camera.py

Remove extra spaces.

* lookup URL when camera is added to hass and add extra guards

balloob committed:
Add stream to the default config (#22602)

bachya committed:
Add permission checking to all RainMachine services (#22399)

* Add permission checking to all RainMachine services

* Linting

* Some initial work

* Owner comments

* Test in place (I think)

* Linting

* Update conftest.py

balloob committed:
Updated frontend to 20190331.0

balloob committed:
Update translations

carstenschroeder committed:
Refactor of ADS integration and introduce ADSEntity (#22583)

* Prevent toogle to false at restart

* change to asyncio.run_coroutine_threadsafe

* refactor ADS platforms; introduce AdsEntity

* fix hound findings

* some formatting

* remove redundant def.

* fix useless super delegation

* fix inconsistent ADS data type for brightness

* fix requested changes

* fix comment

robbiet480 committed:
Add new mobile_app webhook command: get_zones (#22604)

## Description:

Adds a new `mobile_app` webhook command, `get_zones`, which just returns all zones.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

N1nja98 committed:
Fixed brightness reducing after each light change (#22606)

self._brightness max is 255 and hsv brightness max is 100. Assigning 255 based brightness value directly with 100 based hsv reduces brightness eventually to zero.

amelchio committed:
Add netgear_lte connection sensors (#22558)

balloob committed:
Only allow admins to enable remote connection (#22609)

* Only allow admins to enable remote connection

* Protect WS API

* Lint

balloob committed:
Cloudhooks for webhook config flows (#22611)

pvizeli committed:
Hass.io ingress (#22505)

* Fix API stream of snapshot / Add ingress

* fix lint

* Fix stream handling

* Cleanup api handling

* fix typing

* Set proxy header

* Use header constant

* Enable the ingress setup

* fix lint

* Fix name

* Fix tests

* fix lint

* forward params

* Add tests for ingress

* Cleanup cookie handling with aiohttp 3.5

* Add more tests

* Fix tests

* Fix lint

* Fix header handling for steam

* forward header too

* fix lint

* fix flake

awarecan committed:
Config CircleCI workflow (#22590)

* Add mypyrc to control typing check, add mypy to circle

* Add translation upload circlci job

pvizeli committed:
Delete .travis.yml

andrewsayre committed:
Add source selection to Heos component (#22592)

* Add select source support

* Review feedback changes

* Removed unused import

* Ignore 'umused' import used in typing

* Only include trace back on useful errors

* Remove return from play_source

pvizeli committed:
Update face_recognition to 1.2.3 (#22622)

fredrike committed:
Make  platform setup a coroutine (#22620)

* make setup_platform a coroutine

* Update homeassistant/components/tellduslive/sensor.py

Co-Authored-By: fredrike <fredrik.e@gmail.com>

VDRainer committed:
Trend binary sensor check for state unavailable (#22621)

* Trend binary sensor check for state unavailable

Fixes: https://github.com/home-assistant/home-assistant/issues/20210

* Fix pylint

balloob committed:
Disable Z-Wave autoheal (#22628)

Danielhiversen committed:
Handle disonnect bug in Tibber library (#22629)

anna-oake committed:
Fix xiaomi vacuum resume functionality (#22626)

renemarc committed:
Fix GTFS variable type mismatch (#22624)

etheralm committed:
Add support for Dyson Purecool 2018 Air Purifiers models TP04 and DP04 (#22215)

* initial commit

initial commit

rewrite tests

fix merge issue with fan component

fix merge issue with fan component

* correct line length

* change to sync_setup_component for tests

* rename services and move services.yaml

* move hepa and carbon filter state from sensor to fan

* add test for duplicate entities

* fix method call tests

* fix docstring

pnbruckner committed:
Amcrest: Add on/off support & attributes. Bump amcrest to 1.3.0 (#22418)

* Amcrest: Add on/off support & attributes to camera entity. Bump amcrest package to 1.3.0.

Add support for turn_on & turn_off services.

Add implementation of is_recording method, as well as brand, model, hardware_version, machine_name, serial_number, software_build and software_version attributes.

Bump amcrest package to 1.3.0 required for above changes and also handles errors in storage commands which resolves #19982.

* Update per review

Rebase to upstream/dev.

Remove video_enabled property and setter and replace with _enable_video_stream
method.

Remove static attributes from camera and sensors.

exxamalte committed:
change library to georss_generic_client (#22615)

pvizeli committed:
Support GET params for websocket ingress path (#22638)

awarecan committed:
Require static-check success first for rest of workflow (#22635)

* Require static-check success first

* Update config.yml

awarecan committed:
Add trusted networks deprecating warning (#22487)

* Add trusted networks deprecating warning

* Update auth.py

* Update auth.py

* Update auth.py

* Update auth.py

* Tweak

awarecan committed:
Add codecov (#22649)

kellerza committed:
Qwikswitch fix listen loop (#22600)

* Qwikswitch fix listen loop
* 0.93 fix qwikcord upstream

balloob committed:
Run PyLint under Python 3.5 (#22642)

* Run PyLint under Python 3.5

* Remove -q from pip install to debug

* Upgrade setuptools before install

* Use correct cache key for pylint

rohankapoorcom commented on .github/main.workflow:
```diff
@@ -1,41 +0,0 @@
-workflow "Python 3.7 - tox" {
```
Please revert all changes to this file.

gadgetchnnel replied:
That file also appears to have been deleted from the dev branch here: https://github.com/home-assistant/home-assistant/tree/dev/.github
Do I still need to do anything with this?

rohankapoorcom replied:
If you can do a rebase so that it doesn't show any changes to that file (I know it's deleted, but we're trying to keep a clean git history), that would be ideal.

rohankapoorcom commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -199,14 +215,26 @@ def _async_render(self):
+        ))
+
+        if self._attribute_templates is not None:
+            templates.update(dict(('_attributes.' + key, value)
```
Use string formatting instead of concatenation.

rohankapoorcom replied:
Nit: I prefer to use the syntax `{}` rather than `dict()`

rohankapoorcom replied:
Pull `_attributes.` to a const

rohankapoorcom commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -199,14 +215,26 @@ def _async_render(self):
 
             try:
-                setattr(self, property_name, template.async_render())
+                if property_name.startswith("_attributes."):
```
Don't use double quotes when the rest of the file uses single quotes.

rohankapoorcom reviewed (changes requested):
Looks like a good start 😄!

In the interest of making this easier to review and merge, would you mind splitting out the `sensor` platform from the `binary_sensor`? Let's keep the `binary_sensor` in this PR and we can handle the `sensor` in a follow up one (feel free to make WIP one now until this is merged).

Jc2k committed:
Fix racy homekit_controller platform setup caused by #22368 (#22655)

zxdavb committed:
Improve evohome exception handling and fix bugs (#22140)

* Use latest client library, evohomeclient v0.3.1

* Fix issue #22097: Failed to call service climate/turn_on...

* BUGFIX: handle case where a Zone doesn't have a temperature

* BUGFIX: missing exception handler, and inappropriate delint hints

* Improve exception handling, and also better messages

* improve code (REDACT secrets); remove TODOs

* minor refactor - improve error message

* more refactoring - improve error message

* remove TODOs

* update to latest evohomeclient library

* Use latest client library, evohomeclient v0.3.1

* Fix issue #22097: Failed to call service climate/turn_on...

* BUGFIX: handle case where a Zone doesn't have a temperature

* BUGFIX: missing exception handler, and inappropriate delint hints

* Improve exception handling, and also better messages

* improve code (REDACT secrets); remove TODOs

* minor refactor - improve error message

* more refactoring - improve error message

* remove TODOs

* update to latest evohomeclient library

* fix requests for houndci-bot

* Tidy up requests exception handling

* Correct lint error

* update to latest client library

* minor de-lint

* more cleanup of exceptions, messages

* refactored for new exception

* fix error in requirements*_all.txt

* de-lint

* delint unused import

* import 3rd-party library only inside methods

* change honeywell tests

* delint, fix typo

* we dont log usernames, passwords, etc.

* de-lint

Adminiuga committed:
Don't force updates on ZHA Electrical Measurement sensor. (#22647)

cgtobi committed:
Fix pytest durations parameter (#22658)

* Fix durations parameter

* Update config.yml

balloob committed:
Admin service to automatically add empty schema (#22637)

* Admin service to automatically add empty schema

* Lint

gadgetchnnel committed:
Updated binary_sensor

gadgetchnnel committed:
Merge branch 'template-sensor-attributes' of https://github.com/gadgetchnnel/home-assistant into template-sensor-attributes

gadgetchnnel commented:
Looks like this PR has got a little messy (after an attempted rebase). Is it going to be better to close this and create a new one?

rohankapoorcom commented:
Yep, it looks like at the last step instead of force pushing your rebased changes back onto the remote branch, you may have accidentally merged them back in. I think it will be cleaner if you can open a new PR. Please link the old PR to the new one and feel free to tag me to continue reviewing.
